### ISSUE TITLE
post_run_reset in Executors

### ISSUE THREAD

**Issue by anneborcherding** [AFLplusplus/LibAFL#1642]
Hi there,

for me it is not yet intuitive if and when a wrapping Executor calls the `post_run_reset` functions of the Executors it wraps.

Right now, all Executors call the `post_run_reset` method, and as a result, if I am stacking a CombinedExecutor and a WithObserver Executor on top of my own Executor, the `post_run_reset` function is not called as I would expect.

https://github.com/AFLplusplus/LibAFL/blob/745326ee26236e2233e28a9f2f6e006d277a86f3/libafl/src/executors/combined.rs#L60C1-L62C41

Here, we get the following behavior:

``` 
let ret = self.primary.run_target(fuzzer, state, mgr, input); // calls run_target and post_run_reset of wrapped executor, as defined in WithObservers
self.primary.post_run_reset(); // does nothing (default)
self.secondary.post_run_reset(); // does nothing (default)
```

In my understanding, it would be best to differentiate between Executors that wrap a single other Executor (such as the WithObservers Executor), and Executors that combine several Executors. Executors with only one other Executor should not call the `post_run_reset` function in order not to change the underlying Executor's behavior. In contrast, Executors that combine several Executors should be able to combine the Executors in each way they want, including calling the `post_run_reset` function.

However, this all assumes that it is generally okay to stack several Executors.

What are your thoughts on this? I think the easiest approach would be to remove the call to `post_run_reset` from the executors that wrap only one other executor and to implement the `post_run_reset` function for them, which will just call the corresponding function in the wrapped Executor.

**Comment by tokatoka** [comment]
> for me it is not yet intuitive if and when a wrapping Executor calls the post_run_reset functions of the Executors it wraps.

Yes we know it that these APIs are bad and we have a plan to refactor it in near future.
Originally the post_run_reset() was for TimeoutExecutor, but now I think wrapping wrapping executors to implement timeout is a bad choice of design.

 Our plan now is to remove TimeoutExecutor and merge it to other executors. Then no post_run_reset is needed anymore.
@andreafioraldi said he'll work on in December.

**Comment by anneborcherding** [comment]
Allright! Nevertheless, I think a `post_run_reset` would still be good to have, to allow the executors to clean up after running (e.g. running a reset script on the server).

**Comment by domenukk** [comment]
The clean up can just happen on execute, right? Or would you specifically want to have that outside of the evaluation/after the evaluation/async?

**Comment by anneborcherding** [comment]
For my usecase, I am combining several executors that have different tasks and each of the executors gets called to cleanup its specific things after the whole execution (consiting of several executors) is finished.

To be more specific, I have an executor which starts a network sniffing to record the network traffic of the actual network-based execution. This sniffing needs to be stoped after the actual execution is finished. Right now, I am solving this by stopping the network sniffing in the post_run_reset function, but there might be a different solution to this as well.

**Comment by tokatoka** [comment]
can you use observer to do it instead?
You should have something like NetworkSniffingObserver right? You could do the clean there too.

In my opinion, the post_run_reset should be removed
